### PR TITLE
feat(AudioResource)!: use metadata instead of name

### LINKED ISSUE #68: Allow arbitrary metadata to be attached to resources
Currently, you can only pass a string, `name`, as metadata that can be attached to a track. This should be replaced with `metadata`, that allows a user to pass whatever they want in this place.

### PR DESCRIPTION
**Pull request by amishshah**
**Please describe the changes this PR makes and why it should be merged:**

Resolves #68

Example usage:

```ts
interface TrackMetadata {
  title: string;
  artist: string;
}

const resource = createAudioResource<TrackMetadata>('/my/path/file.mp3', {
  metadata: {
    title: 'make discord.js voice great again speech',
    artist: 'hydrabolt'
  }
});
```

**Status and versioning classification:**

- Code changes have been tested against the Discord API, or there are no code changes
- This PR changes the library's interface (methods or parameters added)
- This PR includes breaking changes (methods removed or renamed, parameters moved or removed)

### CODE REVIEW

**Review by papaia — COMMENTED** [review]
I believe defaulting `T` to unknown would be a better approach.
Since the metadata is optional, shouldn't that type argument be optional too?

**Comment by amishshah** [comment]
@papaia @kyranet changes made!

**Comment by papaia** [comment]
Might wanna attach the resource to the error object(s) for easier debugging (in place of the old behavior of including the resource name in the error message).

File: src/audio/AudioResource.ts [hunk 1]
```diff
@@ -7,17 +7,17 @@ import type { AudioPlayer } from './AudioPlayer';
...
-interface CreateAudioResourceOptions {
+interface CreateAudioResourceOptions<T> {
 	/**
 	 * The type of the input stream. Defaults to `StreamType.Arbitrary`.
 	 */
 	inputType?: StreamType;
 
 	/**
-	 * An optional name that can be attached to the resource. This could be a track title, song name etc.
+	 * Optional metadata that can be attached to the resource (e.g. track title, random ID).
```
Comment (papaia) [inline src/audio/AudioResource.ts:17]: Maybe add `@see AudioResource#metadata`?

**Comment by amishshah** [comment]
Thanks for the suggestions @papaia, changes have been made.

File: src/audio/AudioResource.ts [hunk 2]
```diff
@@ -16,6 +16,7 @@ interface CreateAudioResourceOptions<T> {
 	/**
 	 * Optional metadata that can be attached to the resource (e.g. track title, random ID).
 	 * This is useful for identification purposes when the resource is passed around in events.
+	 * @see AudioResource#metadata
```
Comment (iCrawl) [inline src/audio/AudioResource.ts:19]: See is not a keyword used with typedoc for this kind of thing.

https://typedoc.org/guides/doccomments/#symbol-references

File: src/audio/AudioResource.ts [hunk 3]
```diff
@@ -7,17 +7,18 @@ import type { AudioPlayer } from './AudioPlayer';
...
 	/**
 	 * The type of the input stream. Defaults to `StreamType.Arbitrary`.
 	 */
 	inputType?: StreamType;
 
 	/**
-	 * An optional name that can be attached to the resource. This could be a track title, song name etc.
+	 * Optional metadata that can be attached to the resource (e.g. track title, random ID).
 	 * This is useful for identification purposes when the resource is passed around in events.
+	 * @see AudioResource#metadata
```
Comment (papaia) [inline src/audio/AudioResource.ts:19]:
```suggestion
	 * {@link AudioResource.metadata}
```

**Review by kyranet — APPROVED** [review]
Nice

**Pull request merged**